apps/aecore/lib/aecore/util/state_trees.ex

```diff
@@ -0,0 +1,88 @@
+defmodule Aecore.Util.StateTrees do
+  @moduledoc """
+  Abstract module defining functions for known Tree types
+  """
+  defmacro __using__([tree_type, stored_type]) when is_atom(tree_type) do
+    quote location: :keep do
+      alias Aecore.Contract.{Contract, ContractStateTree}
+      alias Aecore.Chain.Identifier
+      alias Aecore.Naming.{Name, NameCommitment}
+      alias Aecore.Util.StateTrees
+      alias Aeutil.PatriciaMerkleTree
+      alias Aeutil.Serialization
+      alias MerklePatriciaTree.Trie
+
+      @typedoc "Hash of the tree"
+      @type hash :: binary()
+
+      @spec init_empty :: Trie.t()
+      def init_empty do
+        PatriciaMerkleTree.new(unquote(tree_type))
+      end
+
+      @spec put(Trie.t(), binary(), map()) :: Trie.t() | {:error, String.t()}
+      def put(tree, key, value) do
+        if StateTrees.valide_store_type?(value, unquote(stored_type)) do
+          serialized_state = Serialization.rlp_encode(value)
+          PatriciaMerkleTree.enter(tree, key, serialized_state)
+        else
+          {:error,
+           "#{__MODULE__}: Invalid value type: #{value.__struct__} but expected value type #{
+             unquote(stored_type)
+           }"}
+        end
+      end
+
+      @spec get(Trie.t(), binary()) :: map() | :none | {:error, String.t()}
+      def get(tree, key) do
+        case PatriciaMerkleTree.lookup(tree, key) do
+          {:ok, serialized_value} ->
+            {:ok, deserialized_value} = Serialization.rlp_decode_anything(serialized_value)
+
+            if StateTrees.valide_store_type?(deserialized_value, unquote(stored_type)) do
+              __MODULE__.process_struct(deserialized_value, key, tree)
```

Comment:
The `process_struct` function is required to be implemented by ***StateTrees. There should be a callback specification that it's required or you can add a default.

Reply:
You are right! You will fix this. Thank you